src/screens/DiscoveryScreenV2/hooks/useDiscovery.js

```diff
@@ -1,13 +1,94 @@
-import {Context} from '../../../context';
 
 const useDiscovery = () => {
   const [, followingDispatch] = React.useContext(Context).following;
   const [refreshing, setRefreshing] = React.useState(false);
+  const [discovery, discoveryDispatch] = React.useContext(Context).discovery;
+
+  const users = React.useMemo(() => {
+    return discovery.initialUsers.map((item) => ({
+      ...item,
+      following: item.following !== undefined ? item.following : item.user_id_follower !== null
+    }));
+  }, [discovery]);
+
+  const exhangeFollower = (newUserLists, willFollow, userId) => {
+    const indexUser = newUserLists.findIndex((item) =>
+      item.user ? item.user.user_id === userId : item.user_id === userId
+    );
+    if (newUserLists[indexUser].user) {
+      newUserLists[indexUser].user.following = !!willFollow;
+      newUserLists[indexUser].user.is_followed = !!willFollow;
+    } else {
+      newUserLists[indexUser].following = !!willFollow;
+      newUserLists[indexUser].is_followed = !!willFollow;
+    }
+    return newUserLists[indexUser];
+  };
+
+  const topicExchangeFollower = (newTopicLists, willFollow, topicId, fromTopicPage) => {
+    let indexTopic;
+    if (fromTopicPage) {
+      indexTopic = newTopicLists.findIndex((item) => item.name === topicId);
+    } else {
+      indexTopic = newTopicLists.findIndex((item) => item.topic_id === topicId);
+    }
+    newTopicLists[indexTopic].following = !!willFollow;
+    return newTopicLists[indexTopic];
+  };
+
+  const mapUser = (newUser) => {
+    return discovery.initialUsers.map((user) => {
+      if (user.user) {
+        if (user.user.user_id === newUser.user.user_id) return newUser;
+      } else if (user.user_id === newUser.user_id) return newUser;
+      return user;
+    });
+  };
+
+  const topics = React.useMemo(() => {
+    return discovery.initialTopics.map((item) => ({
+      ...item,
+      following: item.following !== undefined ? item.following : item.user_id_follower !== null
+    }));
+  }, [discovery.initialTopics]);
+
+  const mapTopic = (newTopic) => {
+    return discovery.initialTopics.map((topic) => {
+      if (topic.topic_id === newTopic.topic_id) {
+        return newTopic;
+      }
+      return topic;
+    });
+  };
+
+  const updateFollowDiscoveryContext = (willFollow, item) => {
+    const newFollowedUsers = [...users];
+    const newUser = exhangeFollower(
+      newFollowedUsers,
+      willFollow,
+      item.user ? item.user.user_id : item.user_id
+    );
+
+    DiscoveryAction.setDiscoveryInitialUsers(mapUser(newUser), discoveryDispatch);
+  };
+
+  const updateFollowTopicDiscoveryContext = (willFollow, item, fromTopicPage) => {
+    const newFollowedTopics = [...topics];
+    const newTopic = topicExchangeFollower(
+      newFollowedTopics,
+      willFollow,
+      item.topic_id,
+      fromTopicPage
+    );
+    DiscoveryAction.setDiscoveryInitialTopics(mapTopic(newTopic), discoveryDispatch);
+  };
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `updateFollowTopicDiscoveryContext` function is similar to `updateFollowDiscoveryContext`, but for topics. It's also a good practice for managing state.



<!-- This is an auto-generated comment by OSS CodeRabbit -->